File: programmers/연습문제/level1/키패드누르기.cpp
/*
    2020 카카오 인턴십 1번 문제였다.
    그 때 쉽다고 생각했는데 답이 제대로 안나와 시간을 많이 잡아먹었었다..
    그런데 다시 풀어보니 쉽게 해결했다.

    먼저 휴대폰을 배열로 생각하고 각 번호의 좌표를 미리 정해놓았다.
    그 후 1, 4, 7이면 무조건 왼쪽
    3, 6, 9면 무조건 오른쪽
    그 외의 숫자는 좌표 차이 계산을 통해 왼손이 더 가까우면 왼손, 오른손이 더 가까우면 오른손으로 설정해주었고
    같은 경우 왼손잡이는 왼쪽, 오른손 잡이는 오른쪽으로 정해주었다.

    풀이를 비교해보니 방식은 비슷하게 푼 것 같다.
*/
#include <string>
#include <vector>

using namespace std;

pair<int, int> phone[]= {
    {3, 1},
    {0, 0},
    {0, 1}, 
    {0, 2},
    {1, 0},
    {1, 1},
    {1, 2},
    {2, 0},
    {2, 1},
    {2, 2}
};

string solution(vector<int> numbers, string hand) {
    string answer = "";
    pair<int, int> left=make_pair(3, 0);
    pair<int, int> right=make_pair(3, 2);
    
    for(int i=0;i<numbers.size();i++) {
        if(numbers[i]==1 || numbers[i]==4 || numbers[i]==7) {
            left=phone[numbers[i]];
            answer+='L';
        } else if(numbers[i]==3 || numbers[i]==6 || numbers[i]==9) {
            right=phone[numbers[i]];
            answer+='R';
        } else {
            int l=abs(left.first-phone[numbers[i]].first) + abs(left.second-phone[numbers[i]].second);
             int r=abs(right.first-phone[numbers[i]].first) + abs(right.second-phone[numbers[i]].second);
            
            if(l<r) {
                left=phone[numbers[i]];
                answer+='L';
            } else if(l>r) {
                right=phone[numbers[i]];
                answer+='R';
            } else if(l==r) {
                if(hand=="left") {
                    left=phone[numbers[i]];
                    answer+='L';
                } else if(hand=="right") {
                    right=phone[numbers[i]];
                    answer+='R';
                }
            }
        }
    }
    
    
    return answer;
}

File: kakao_java/intern2020/키패드누르기.java
package intern2020;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class 키패드누르기 {
    private Hand left;
    private Hand right;
    private Set<Integer> leftHands = new HashSet<>(List.of(1, 4, 7));
    private Set<Integer> rightHands = new HashSet<>(List.of(3, 6, 9));

    private int[] x = {3, 0, 0, 0, 1, 1, 1, 2, 2, 2};
    private int[] y = {1, 0, 1, 2, 0, 1, 2, 0, 1, 2};

    private StringBuilder sb = new StringBuilder();

    public String solution(int[] numbers, String hand) {
        this.left = new Hand(3, 0, 'L');
        this.right = new Hand(3, 2, 'R');

        for (int num : numbers) {
            if (leftHands.contains(num)) {
                sb.append(left.type);
                left.move(x[num], y[num]);
            } else if (rightHands.contains(num)) {
                sb.append(right.type);
                right.move(x[num], y[num]);
            } else {
                appendResult(hand, x[num], y[num]);
            }
        }

        return sb.toString();
    }

    private void appendResult(String hand, int x, int y) {
        int ld = getDistance(left, x, y);
        int rd = getDistance(right, x, y);
        if (ld > rd) {
            sb.append(right.type);
            right.move(x, y);
        } else if (ld < rd) {
            sb.append(left.type);
            left.move(x, y);
        } else {
            if ("right".equals(hand)) {
                sb.append(right.type);
                right.move(x, y);
            } else {
                sb.append(left.type);
                left.move(x, y);
            }
        }
    }

    private int getDistance(Hand hand, int x, int y) {
        return Math.abs(hand.x - x) + Math.abs(hand.y - y);
    }

    static class Hand {
        int x;
        int y;
        char type;

        public Hand(int x, int y, char type) {
            this.x = x;
            this.y = y;
            this.type = type;
        }

        public void move(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}
